racket/collects/racket/hash-code.rkt

```diff
@@ -0,0 +1,46 @@
+#lang racket/base
+
+(provide mix-hash-code hash-code-combine hash-code-combine-unordered)
+
+(require racket/fixnum)
+
+;; unsigned integer view of -2, not a fixnum
+(define (um2) (arithmetic-shift (most-positive-fixnum) 1))
+
+;; fixnum shift-left logical with wraparound
+(define fxsll/wraparound fxlshift/wraparound)
+
+;; fixnum logical exclusive or
+(define fxlogxor fxxor)
+
+;; fixnum shift-right logical
+;; shifting negative fixnums can produce large positive fixnums
```

Comment:
I don't see how that can be the case. It seems like an arithmetic shift should always preserve the sign of a negative fixnum. What am I missing?

Reply:
Oh, I'm missing that the "l" means logical!

Reply:
Still, the result is a fixnum, so why is special treatment needed? (Sorry - intended to ask long ago, but apparently left the comment as "pending".)

Reply:
Is there a better way to do a fixnum shift-right-logical that doesn't require converting a negative fixnum to a large positive big-integer that can be shifted-right back into a fixnum?

Reply:
I've updated the `fxsrl` function to use a different method of calculating the logical right-shift when `a` is negative:
```racket
     (fx-/wraparound (fxrshift (bitwise-and a (most-positive-fixnum)) b)
                     (fxrshift (most-negative-fixnum) b))
```
This should do the same thing without needing to convert it to a large positive big-integer and back.

Reply:
The question I should have asked is "why use a logical shift instead of an arithmetic shift?". The answer might be "because hash-code.ss uses a logical shift". But should hash-code.ss use an arithmetic shift, instead? Does it matter either way?

Reply:
A logical right-shift can preserve a distinction between negative numbers that would otherwise all get shifted to the same number, `-1`, by an arithmetic right-shift.

Reply:
Sorry that I'm being so dense here, but I still don't get it. What two negative numbers when shifted by 10 would produce the same result with sra but have different results for srl?

Reply:
For them to be different they need to be shifted by different amounts. If one is shifted by 10 and the other by 9, they'll be different after a logical shift but the same `-1` after an arithmetic shift.

Reply:
So, it doesn't lose information in a hash-code calculation, since the function is only ever called with 10 as a second argument? It still seems like the readily available fixnum operation would work here.